File: DS_CP_AT/Two Pointers/Reverse Words in a String/reverseWords.rb
=begin
  Given a sentence, reverse the order of its words without affecting the order of letters
  within a given word.
=end

def reverseWords(sentence)

  sentence = sentence.strip
  sentence.gsub(/  +/, " ")

  split_sentence = sentence.split(' ')

  start = 0
  last = split_sentence.length - 1
  temp = ''

  while start < last
    temp = split_sentence[start]
    split_sentence[start] = split_sentence[last]
    split_sentence[last] = temp
    start += 1
    last -= 1
  end

  split_sentence.join(' ')

end

puts reverseWords("We love Python") == "Python love We"
puts reverseWords("To be or not to be") == "be to not or be To"
puts reverseWords("You are amazing") == "amazing are You"
puts reverseWords(" Hello     World ") == "World Hello"
puts reverseWords("Hey") == "Hey"

File: DS_CP_AT/Two Pointers/Reverse Words in a String/ReverseWords.java
public class ReverseWords {

  public static String reverseWords (String sentence) {

    sentence = sentence.trim().replaceAll("\\s+", " ");

    String [] words = sentence.split(" ");

    int back = 0;
    int front = words.length - 1;
    String temp = " ";

    while (back < front) {
      temp = words[back];
      words[back] = words[front];
      words[front] = temp;
      back ++;
      front --;
    }

    return String.join(" ", words);
  }

  public static void main(String[] args) {

    System.out.println(reverseWords("We love Python"));
    System.out.println(reverseWords("To be or not to be"));
    System.out.println(reverseWords("You are amazing"));
    System.out.println(reverseWords(" Hello     World "));
    System.out.println(reverseWords("Hey"));

  }
}
